interfaces/builtin/utils.go

```diff
@@ -110,7 +111,20 @@ func aareExclusivePatterns(orig string) []string {
 // dirs.SnapDesktopFilesDir, but explicitly denies access to all other snaps'
 // desktop files since xdg libraries may try to read all the desktop files
 // in the dir, causing excessive noise. (LP: #1868051)
-func getDesktopFileRules(snapInstanceName string) []string {
+func getDesktopFileRules(snapInstanceName string, spec *apparmor.Specification) []string {
+	// desktop-launch allows to read all .desktop files; but "deny" rules overrule any "allow"
+	// rule, so we must not add these rules if this snap uses the desktop-launch interface.
+	// Also, for security reasons, all these rules are removed if the desktop-launch interface
+	// is listed, thus only if it is really connected will the snap have any kind of access to
+	// these folders/files.
+	if spec != nil {
+		for _, plug := range spec.SnapAppSet().Info().Plugs {
```

Comment:
this doesn't quite mean that the interface is actually connected

Reply:
the cleaner solution would be to think through some sort of optional conflict mechanism for snippets coming from different interfaces

Reply:
I know that it doesn't mean that the interface is actually connected, but in this case that's not a problem because the snippet added just gives more privileges (allows to read the list of files at the DESKTOP folders), so if it isn't added because the other interface is requested, then the snap just will have less privileges because it won't be able to read anything from those folders.

Anyway, I agree in that it's an ugly way of doing things. About how to think about the conflicting mechanism... the problem that I see is that doing it very generic would require a parser for apparmor, which clearly is something excessive. The only idea that I think that could be implemented in that way is to identify blocks of snippets with some kind of unique ID, and that each interface that tries to implement that block will specify a priority, thus only the block with the highest priority would be finally added.

The other solution could be to just implement a REST API call that allows to read the contents of a .desktop file...

Reply:
FWIW as part of new development work AppArmor is adding support for rule priorities itself - so this could eventually be used as a solution here but currently it is not available so perhaps in snapd we need to encode priorities for snippets as suggested. (see https://docs.google.com/presentation/d/13KMOgK7NPvWO51WdwEjik2HANTTU_drdjID7CwrQkdQ/edit#slide=id.g2c72fb4f079_0_56 as an example) (cc @jrjohansen)

Reply:
base priorities are ready now. There could be a few more tests. What is missing is
- logprof/genprof support, which isn't needed for snaps
- priority blocks, which might be nice to have for snippets. It depends on if you want to apply priority to the a whole block of rules in one go.

The priority blocks are something that might happen this week, as that is code that is getting shared with other important work.

Reply:
Mmm... but is that compatible with old kernels? If using priorities break compatibility with old versions... (or am I missing something?)

Reply:
It is compatible with old kernels. The compiler resolves the priorities as part of the compile.

Reply:
Oh! That makes sense. Thanks!